searchkit/constraints.py

```diff
@@ -109,578 +107,691 @@ def __repr__(self):
+
+    # Number of lines to search forwards when the algorithm encounters lines
+    # with no date.
+    MAX_TRY_FIND_WITH_DATE_ATTEMPTS = 500
+
+    LINE_FEED_TOKEN = b'\n'
+
+    def __init__(self, fd, c):
+        self.file = fd
+        self.constraint = c
+        self.line_info = None
+        self.found_any_date = False
+        self.lookup_times = 0
+        with NonDestructiveFileRead(self.file) as f:
+            self.length = f.seek(0, 2)
+
+    def find_token_reverse(self, start_offset, horizon,
+                           attempts=MAX_SEEK_HORIZON_EXPAND):
+        r"""Find `token` in `file` starting from `start_offset` and backing off
+        `horizon`bytes on each iteration for maximum of `max_iterations` times.
+
+        @param start_offset (int): start offset of search
+        @param horizon (int): Amount of bytes to be processed on each step.
+        @param attempts (int, optional): Maximum amount of search iterations.
+                                         Defaults to MAX_SEEK_HORIZON_EXPAND.
+
+        @return:
+            FindDelimiterResult(FindTokenStatus.FOUND, ...) if token is found
+            FindDelimiterResult(FindTokenStatus.REACHED_EOF, ...) if token is
+                not found because the scan reached the EOF
+            FindDelimiterResult(FindTokenStatus.FAILED, ...) if token is
+                not found because scan exhausted `max_iterations`
+        """
+
+        current_offset = -horizon
+        while True:
+            attempts -= 1
+            read_offset = start_offset + current_offset
+            read_offset = read_offset if read_offset > 0 else 0
+            read_size = horizon
+            if start_offset + current_offset <= 0:
+                read_size = read_size + (start_offset + current_offset)
+
+            self.file.seek(read_offset)
+            chunk = self.file.read(read_size)
+            if not chunk or len(chunk) == 0:
+                # We've reached the start of the file and could not find the
+                # token.
+                return SearchState(status=FindTokenStatus.REACHED_EOF,
+                                   offset=0)
+
+            chunk_offset = chunk.rfind(self.LINE_FEED_TOKEN)
+
+            if chunk_offset != -1:
+                return SearchState(status=FindTokenStatus.FOUND,
+                                   offset=read_offset + chunk_offset)
+
+            if attempts <= 0:
+                break
 
-        @param pos: position in a file.
-        """
-        if self._line_date_is_valid(datetime_obj):
-            return self.fd_info.fd.tell()
+            current_offset = current_offset - len(chunk)
+            if (start_offset + current_offset) < 0:
+                return SearchState(status=FindTokenStatus.REACHED_EOF,
+                                   offset=0)
+
+        return SearchState(FindTokenStatus.FAILED)
+
+    def find_token(self, start_offset, horizon,
+                   attempts=MAX_SEEK_HORIZON_EXPAND):
```

Comment:
Same comment here as with find_token_reverse i.e.:

"attempts" kwarg is not actually provided by any caller atm so can we remove it and just set attempts = MAX_SEEK_HORIZON_EXPAND inside the method instead.

Reply:
done